src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Endpoints/ProjectContexts.cs

```diff
@@ -1,51 +1,62 @@
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
 using Microsoft.VisualStudio.LanguageServer.Protocol;
-using Microsoft.VisualStudio.LanguageServerClient.Razor.Extensions;
 using StreamJsonRpc;
 
 namespace Microsoft.VisualStudio.LanguageServerClient.Razor;
 
 internal partial class RazorCustomMessageTarget
 {
     [JsonRpcMethod(CustomMessageNames.RazorProjectContextsEndpoint, UseSingleObjectParameterDeserialization = true)]
-    public async Task<VSProjectContextList?> ProjectContextsAsync(DelegatedProjectContextsParams request, CancellationToken cancellationToken)
+    public Task<VSProjectContextList?> ProjectContextsAsync(DelegatedProjectContextsParams request, CancellationToken _)
     {
-        var hostDocument = request.Identifier.TextDocumentIdentifier;
-        var (synchronized, virtualDocument) = await _documentSynchronizer.TrySynchronizeVirtualDocumentAsync<CSharpVirtualDocumentSnapshot>(
-            _documentManager,
-            request.Identifier.Version,
-            hostDocument,
-            cancellationToken).ConfigureAwait(false);
-
-        if (!synchronized)
-        {
-            return null;
-        }
+        // Previously we would have asked Roslyn for their ProjectContexts, so we can make sure we pass a ProjectContext they understand
+        // to them when we ask them for things. When we generate unique file names for generated files, we no longer need to do that
+        // as the generated file will only be in one project, so we can just use our own ProjectContexts. This makes other things much
+        // easier because we're not trying to understand Roslyn concepts.
+
+        var projects = _projectSnapshotManagerAccessor.Instance.GetProjects();
 
-        var projectContextParams = new VSGetProjectContextsParams()
+        using var projectContexts = new PooledArrayBuilder<VSProjectContext>(capacity: projects.Length);
+
+        var documentFilePath = DocumentFilePathProvider.GetProjectSystemFilePath(request.Uri);
+
+        foreach (var project in projects)
         {
-            TextDocument = new TextDocumentItem()
+            if (project is ProjectSnapshot snapshot &&
+                project.GetDocument(documentFilePath) is not null)
             {
-                LanguageId = CodeAnalysis.LanguageNames.CSharp,
-                Uri = virtualDocument.Uri,
-                Version = virtualDocument.Snapshot.Version.VersionNumber,
-                Text = virtualDocument.Snapshot.GetText(),
+                var projectFileName = Path.GetFileNameWithoutExtension(snapshot.HostProject.FilePath);
+                projectContexts.Add(new VSProjectContext
+                {
+                    Id = project.Key.Id,
+                    Kind = VSProjectKind.CSharp,
+                    Label = snapshot.HostProject.DisplayName is { } displayName
+                        ? $"{projectFileName} ({displayName})"
+                        : projectFileName
```

Comment:
Sadly I can't put up a screenshot of the navbar, but this is what the `projectContexts` list looks like for a multi-targeted project:

![image](https://github.com/dotnet/razor/assets/754264/bd423d63-9624-4acc-bfc5-42def3fad7e3)